File: ios/Tradeguruelectrical/ViewModels/ChatEngine.swift
import Foundation
@preconcurrency import AVFoundation

@Observable
@MainActor
class ChatEngine {
    var isStreaming = false
    var streamingBlocks: [ContentBlock] = []
    var error: String?
    var lastResponseId: String?
    var pipelineStage: PipelineStage = .idle

    private var lastSendMode: ThinkingMode = .faultFinder
    private var lastSendAttachments: [MessageAttachment]?
    private var lastSendType: SendType = .text
    private enum SendType { case text, vision, document }

    private var deviceId: String
    private var audioPlayer: AVAudioPlayer?

    private var currentJWT: String? {
        AuthManager.shared.tokens?.accessToken
    }

    init() {
        deviceId = DeviceManager.deviceIdOrFallback()
        Task { await registerDeviceIfNeeded() }
    }

    func fetchResponse(mode: ThinkingMode, conversation: Conversation) async {
        #if DEBUG
        if APIConfig.useMockData {
            await mockResponse(mode: mode, conversation: conversation)
            return
        }
        #endif

        let apiMessages = conversation.messages.map { msg -> [String: String] in
            let content = msg.blocks.compactMap(\.content).joined(separator: "\n")
            return ["role": msg.role.rawValue, "content": content]
        }

        isStreaming = true
        streamingBlocks = []
        error = nil
        lastSendMode = mode
        lastSendAttachments = nil
        lastSendType = .text

        for await result in TradeGuruAPI.chat(messages: apiMessages, mode: mode, deviceId: deviceId, jwt: currentJWT) {
            switch result {
            case .block(let block):
                streamingBlocks.append(block)
            case .status(let payload):
                if let stage = PipelineStage(rawValue: payload.stage) {
                    pipelineStage = stage
                }
            case .done(let payload):
                lastResponseId = payload.responseId
                finalizeResponse(mode: mode, conversation: conversation)
            case .error(let payload):
                handleStreamError(payload: payload, mode: mode, conversation: conversation)
            }
        }

        if isStreaming { isStreaming = false }
    }

    func fetchVisionResponse(mode: ThinkingMode, attachments: [MessageAttachment]?, conversation: Conversation) async {
        #if DEBUG
        if APIConfig.useMockData {
            await mockResponse(mode: mode, conversation: conversation)
            return
        }
        #endif

        let lastUserMsg = conversation.messages.last { $0.role == .user }
        let messageText = lastUserMsg?.blocks.compactMap(\.content).joined(separator: "\n") ?? ""

        var imageDataUri = ""
        if let atts = attachments {
            for att in atts {
                if att.type == .image, let data = att.thumbnailData {
                    imageDataUri = "data:image/jpeg;base64,\(data.base64EncodedString())"
                    break
                }
            }
        }

        isStreaming = true
        streamingBlocks = []
        error = nil
        lastSendMode = mode
        lastSendAttachments = attachments
        lastSendType = .vision

        for await result in TradeGuruAPI.chatVision(message: messageText, image: imageDataUri, mode: mode, deviceId: deviceId, jwt: currentJWT) {
            switch result {
            case .block(let block):
                streamingBlocks.append(block)
            case .status(let payload):
                if let stage = PipelineStage(rawValue: payload.stage) {
                    pipelineStage = stage
                }
            case .done(let payload):
                lastResponseId = payload.responseId
                finalizeResponse(mode: mode, conversation: conversation)
            case .error(let payload):
                handleStreamError(payload: payload, mode: mode, conversation: conversation)
            }
        }

        if isStreaming { isStreaming = false }
    }

    func uploadDocumentAndChat(mode: ThinkingMode, attachments: [MessageAttachment]?, conversation: Conversation) async {
        #if DEBUG
        if APIConfig.useMockData {
            await mockResponse(mode: mode, conversation: conversation)
            return
        }
        #endif

        lastSendMode = mode
        lastSendAttachments = attachments
        lastSendType = .document

        if let doc = attachments?.first(where: { $0.type == .document }), let data = doc.thumbnailData {
            let mimeType: String
            if doc.fileName.hasSuffix(".pdf") {
                mimeType = "application/pdf"
            } else if doc.fileName.hasSuffix(".txt") {
                mimeType = "text/plain"
            } else {
                mimeType = "application/octet-stream"
            }
            do {
                let uploaded = try await TradeGuruAPI.uploadFile(
                    fileData: data,
                    fileName: doc.fileName,
                    mimeType: mimeType,
                    deviceId: deviceId,
                    jwt: currentJWT
                )
                let lastUserText = conversation.messages.last(where: { $0.role == .user })?.blocks.compactMap(\.content).joined(separator: "\n") ?? ""
                let messageWithFile = lastUserText.isEmpty
                    ? "[Uploaded: \(uploaded.filename)]"
                    : "\(lastUserText)\n\n[Attached file: \(uploaded.filename) (id: \(uploaded.id))]"

                let apiMessages: [[String: String]] = conversation.messages.map { msg in
                    var content = msg.blocks.compactMap(\.content).joined(separator: "\n")
                    if msg.id == conversation.messages.last?.id && msg.role == .user {
                        content = messageWithFile
                    }
                    return ["role": msg.role.rawValue, "content": content]
                }

                isStreaming = true
                streamingBlocks = []
                error = nil

                for await result in TradeGuruAPI.chat(messages: apiMessages, mode: mode, deviceId: deviceId, jwt: currentJWT) {
                    switch result {
                    case .block(let block):
                        streamingBlocks.append(block)
                    case .status(let payload):
                        if let stage = PipelineStage(rawValue: payload.stage) {
                            pipelineStage = stage
                        }
                    case .done(let payload):
                        lastResponseId = payload.responseId
                        finalizeResponse(mode: mode, conversation: conversation)
                    case .error(let payload):
                        error = payload.message
                        streamingBlocks = []
                        isStreaming = false
                        pipelineStage = .idle
                    }
                }
                if isStreaming { isStreaming = false }
            } catch {
                self.error = "File upload failed: \(error.localizedDescription)"
            }
        } else {
            await fetchResponse(mode: mode, conversation: conversation)
        }
    }

    func retryLastRequest(conversation: Conversation) {
        error = nil
        switch lastSendType {
        case .text:
            Task { await fetchResponse(mode: lastSendMode, conversation: conversation) }
        case .vision:
            Task { await fetchVisionResponse(mode: lastSendMode, attachments: lastSendAttachments, conversation: conversation) }
        case .document:
            Task { await uploadDocumentAndChat(mode: lastSendMode, attachments: lastSendAttachments, conversation: conversation) }
        }
    }

    func rateLastResponse(stars: Int, comment: String? = nil, mode: ThinkingMode) async {
        guard let responseId = lastResponseId else { return }
        try? await TradeGuruAPI.rate(
            responseId: responseId,
            stars: stars,
            mode: mode,
            comment: comment,
            deviceId: deviceId,
            jwt: currentJWT
        )
    }

    func transcribeAudio(_ audioData: Data) async -> String? {
        do {
            return try await TradeGuruAPI.transcribe(audioData: audioData, mimeType: "audio/m4a", deviceId: deviceId, jwt: currentJWT)
        } catch {
            self.error = "Transcription failed"
            return nil
        }
    }

    func speakText(_ text: String) async {
        do {
            let audioData = try await TradeGuruAPI.speak(text: text, deviceId: deviceId, jwt: currentJWT)
            let player = try AVAudioPlayer(data: audioData)
            player.prepareToPlay()
            player.play()
            audioPlayer = player
        } catch {
            self.error = "Speech playback failed"
        }
    }

    private func finalizeResponse(mode: ThinkingMode, conversation: Conversation) {
        let assistantMessage = ChatMessage(role: .assistant, blocks: streamingBlocks, mode: mode)
        conversation.messages.append(assistantMessage)
        conversation.updatedAt = Date()
        if conversation.messages.count == 2 {
            let firstUserText = conversation.messages.first?.blocks.first?.content ?? "Chat"
            conversation.title = String(firstUserText.prefix(40))
        }
        streamingBlocks = []
        isStreaming = false
        pipelineStage = .idle
    }

    private func handleStreamError(payload: StreamErrorPayload, mode: ThinkingMode, conversation: Conversation) {
        error = payload.message
        if payload.partial == true, !streamingBlocks.isEmpty {
            let partialMessage = ChatMessage(role: .assistant, blocks: streamingBlocks, mode: mode)
            conversation.messages.append(partialMessage)
        }
        streamingBlocks = []
        isStreaming = false
        pipelineStage = .idle
    }

    private func registerDeviceIfNeeded() async {
        if deviceId.isEmpty || deviceId == "pending" {
            do {
                let registeredId = try await TradeGuruAPI.registerDevice()
                DeviceManager.save(registeredId)
                deviceId = registeredId
            } catch {
                self.error = "Device registration failed. Retrying on next launch."
            }
        }
    }

    #if DEBUG
    private func mockResponse(mode: ThinkingMode, conversation: Conversation) async {
        isStreaming = true
        streamingBlocks = []
        error = nil

        let mockConversation = MockData.allConversations.first { $0.mode == mode }
        let mockBlocks = mockConversation?.messages.last(where: { $0.role == .assistant })?.blocks ?? [
            ContentBlock(type: .text, content: "This is a mock response for \(mode.name) mode.")
        ]

        try? await Task.sleep(for: .milliseconds(500))

        for block in mockBlocks {
            streamingBlocks.append(block)
            try? await Task.sleep(for: .milliseconds(100))
        }

        finalizeResponse(mode: mode, conversation: conversation)
    }
    #endif
}

File: android/app/src/main/java/com/tradeguru/electrical/viewmodels/ChatEngine.kt
package com.tradeguru.electrical.viewmodels

import android.media.MediaPlayer
import com.tradeguru.electrical.data.DomainMappers
import com.tradeguru.electrical.models.MessageRole
import com.tradeguru.electrical.models.PipelineStage
import com.tradeguru.electrical.models.ThinkingMode
import com.tradeguru.electrical.services.ApiMessage
import com.tradeguru.electrical.services.AuthManager
import com.tradeguru.electrical.services.DeviceManager
import com.tradeguru.electrical.services.StreamResult
import com.tradeguru.electrical.services.TradeGuruAPI
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.File
import java.util.UUID

class ChatEngine(
    private val deviceManager: DeviceManager,
    private val authManager: AuthManager
) {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)

    private var deviceId: String = deviceManager.getOrCreateDeviceId()
    private val currentJwt: String? get() = authManager.currentJwt

    private val _isStreaming = MutableStateFlow(false)
    val isStreaming: StateFlow<Boolean> = _isStreaming

    private val _streamingBlocks = MutableStateFlow<List<DomainMappers.ContentBlock>>(emptyList())
    val streamingBlocks: StateFlow<List<DomainMappers.ContentBlock>> = _streamingBlocks

    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error

    private val _pipelineStage = MutableStateFlow(PipelineStage.IDLE)
    val pipelineStage: StateFlow<PipelineStage> = _pipelineStage

    var lastResponseId: String? = null
        private set

    private enum class SendType { TEXT, VISION, DOCUMENT }

    private var lastSendType = SendType.TEXT
    private var lastSendMode = ThinkingMode.FAULT_FINDER
    private var lastAttachments: List<DomainMappers.MessageAttachment>? = null

    private var mediaPlayer: MediaPlayer? = null

    init {
        registerDeviceIfNeeded()
    }

    suspend fun fetchResponse(
        mode: ThinkingMode,
        conversation: DomainMappers.Conversation
    ) {
        val apiMessages = buildApiMessages(conversation)

        _isStreaming.value = true
        _streamingBlocks.value = emptyList()
        _error.value = null
        lastSendMode = mode
        lastAttachments = null
        lastSendType = SendType.TEXT

        TradeGuruAPI.chat(
            messages = apiMessages,
            mode = mode,
            deviceId = deviceId,
            jwt = currentJwt
        ).collect { result ->
            processStreamEvent(result)
        }

        if (_isStreaming.value) _isStreaming.value = false
    }

    suspend fun fetchVisionResponse(
        mode: ThinkingMode,
        attachments: List<DomainMappers.MessageAttachment>?,
        conversation: DomainMappers.Conversation
    ) {
        val imageAttachment = attachments?.firstOrNull {
            it.type == com.tradeguru.electrical.models.AttachmentType.IMAGE && it.thumbnailData != null
        }

        if (imageAttachment == null || imageAttachment.thumbnailData == null) {
            fetchResponse(mode, conversation)
            return
        }

        val lastUserMsg = conversation.messages.lastOrNull { it.role == MessageRole.USER }
        val messageText = lastUserMsg?.blocks
            ?.mapNotNull { it.content }
            ?.joinToString("\n") ?: ""

        val imageDataUri = try {
            "data:image/jpeg;base64,${
                android.util.Base64.encodeToString(imageAttachment.thumbnailData, android.util.Base64.NO_WRAP)
            }"
        } catch (e: Exception) {
            _error.value = "Failed to encode image"
            return
        }

        if (imageDataUri.length < 30) {
            fetchResponse(mode, conversation)
            return
        }

        _isStreaming.value = true
        _streamingBlocks.value = emptyList()
        _error.value = null
        lastSendMode = mode
        lastAttachments = attachments
        lastSendType = SendType.VISION

        TradeGuruAPI.chatVision(
            message = messageText,
            image = imageDataUri,
            mode = mode,
            deviceId = deviceId,
            jwt = currentJwt
        ).collect { result ->
            processStreamEvent(result)
        }

        if (_isStreaming.value) _isStreaming.value = false
    }

    suspend fun uploadDocumentAndChat(
        mode: ThinkingMode,
        attachments: List<DomainMappers.MessageAttachment>?,
        conversation: DomainMappers.Conversation
    ) {
        lastSendMode = mode
        lastAttachments = attachments
        lastSendType = SendType.DOCUMENT

        val doc = attachments?.firstOrNull {
            it.type == com.tradeguru.electrical.models.AttachmentType.DOCUMENT
        }

        if (doc == null || doc.thumbnailData == null) {
            fetchResponse(mode, conversation)
            return
        }

        val mimeType = when {
            doc.fileName.endsWith(".pdf") -> "application/pdf"
            doc.fileName.endsWith(".txt") -> "text/plain"
            else -> "application/octet-stream"
        }

        try {
            val uploaded = TradeGuruAPI.uploadFile(
                fileData = doc.thumbnailData,
                fileName = doc.fileName,
                mimeType = mimeType,
                deviceId = deviceId,
                jwt = currentJwt
            )

            val lastUserText = conversation.messages
                .lastOrNull { it.role == MessageRole.USER }
                ?.blocks?.mapNotNull { it.content }
                ?.joinToString("\n") ?: ""

            val messageWithFile = if (lastUserText.isEmpty()) {
                "[Uploaded: ${uploaded.filename}]"
            } else {
                "$lastUserText\n\n[Attached file: ${uploaded.filename} (id: ${uploaded.id})]"
            }

            val apiMessages = conversation.messages.map { msg ->
                var content = msg.blocks.mapNotNull { it.content }.joinToString("\n")
                if (msg.id == conversation.messages.lastOrNull()?.id && msg.role == MessageRole.USER) {
                    content = messageWithFile
                }
                ApiMessage(role = msg.role.value, content = content)
            }

            _isStreaming.value = true
            _streamingBlocks.value = emptyList()
            _error.value = null

            TradeGuruAPI.chat(
                messages = apiMessages,
                mode = mode,
                deviceId = deviceId,
                jwt = currentJwt
            ).collect { result ->
                processStreamEvent(result)
            }

            if (_isStreaming.value) _isStreaming.value = false
        } catch (e: Exception) {
            _error.value = "File upload failed: ${e.message}"
        }
    }

    fun retryLastRequest(conversation: DomainMappers.Conversation) {
        _error.value = null
        when (lastSendType) {
            SendType.TEXT -> scope.launch {
                fetchResponse(lastSendMode, conversation)
            }
            SendType.VISION -> scope.launch {
                fetchVisionResponse(lastSendMode, lastAttachments, conversation)
            }
            SendType.DOCUMENT -> scope.launch {
                uploadDocumentAndChat(lastSendMode, lastAttachments, conversation)
            }
        }
    }

    suspend fun rateLastResponse(stars: Int, comment: String?, mode: ThinkingMode) {
        val responseId = lastResponseId ?: return
        withContext(Dispatchers.IO) {
            try {
                TradeGuruAPI.rate(
                    responseId = responseId,
                    stars = stars,
                    mode = mode,
                    deviceId = deviceId,
                    comment = comment,
                    jwt = currentJwt
                )
            } catch (_: Exception) { }
        }
    }

    suspend fun transcribeAudio(audioFile: File): String? {
        return try {
            TradeGuruAPI.transcribe(
                audioData = audioFile.readBytes(),
                mimeType = "audio/m4a",
                deviceId = deviceId,
                jwt = currentJwt
            )
        } catch (_: Exception) {
            _error.value = "Transcription failed"
            null
        }
    }

    suspend fun speakText(text: String) {
        try {
            val audioData = TradeGuruAPI.speak(
                text = text,
                deviceId = deviceId,
                jwt = currentJwt
            )
            mediaPlayer?.release()
            val tempFile = File.createTempFile("tts_", ".mp3")
            tempFile.writeBytes(audioData)
            mediaPlayer = MediaPlayer().apply {
                setDataSource(tempFile.absolutePath)
                prepare()
                start()
                setOnCompletionListener { mp ->
                    mp.release()
                    tempFile.delete()
                }
            }
        } catch (_: Exception) {
            _error.value = "Speech playback failed"
        }
    }

    fun registerDeviceIfNeeded() {
        if (deviceId.isEmpty() || deviceId == "pending") {
            scope.launch {
                try {
                    val newDeviceId = withContext(Dispatchers.IO) {
                        TradeGuruAPI.registerDevice()
                    }
                    deviceManager.getOrCreateDeviceId()
                    deviceId = newDeviceId
                } catch (_: Exception) {
                    _error.value = "Device registration failed. Retrying on next launch."
                }
            }
        }
    }

    fun clearError() {
        _error.value = null
    }

    fun finalizeResponse(
        mode: ThinkingMode,
        conversation: DomainMappers.Conversation
    ): DomainMappers.ChatMessage {
        val blocks = _streamingBlocks.value.toList()
        val assistantMessage = DomainMappers.ChatMessage(
            id = UUID.randomUUID().toString(),
            role = MessageRole.ASSISTANT,
            blocks = blocks,
            timestamp = System.currentTimeMillis(),
            mode = mode
        )

        _streamingBlocks.value = emptyList()
        _isStreaming.value = false
        _pipelineStage.value = PipelineStage.IDLE

        return assistantMessage
    }

    fun handleStreamError(
        error: StreamResult.Error,
        mode: ThinkingMode,
        conversation: DomainMappers.Conversation
    ): DomainMappers.ChatMessage? {
        _error.value = error.payload.message
        var partialMessage: DomainMappers.ChatMessage? = null

        if (error.payload.partial == true && _streamingBlocks.value.isNotEmpty()) {
            partialMessage = DomainMappers.ChatMessage(
                id = UUID.randomUUID().toString(),
                role = MessageRole.ASSISTANT,
                blocks = _streamingBlocks.value.toList(),
                timestamp = System.currentTimeMillis(),
                mode = mode
            )
        }

        _streamingBlocks.value = emptyList()
        _isStreaming.value = false
        _pipelineStage.value = PipelineStage.IDLE

        return partialMessage
    }

    fun resetForNewConversation() {
        _streamingBlocks.value = emptyList()
        _isStreaming.value = false
        _error.value = null
    }

    private fun buildApiMessages(
        conversation: DomainMappers.Conversation
    ): List<ApiMessage> =
        conversation.messages.map { msg ->
            val content = msg.blocks.mapNotNull { it.content }.joinToString("\n")
            ApiMessage(role = msg.role.value, content = content)
        }

    private fun processStreamEvent(result: StreamResult) {
        when (result) {
            is StreamResult.Block -> {
                _streamingBlocks.value = _streamingBlocks.value + result.block
            }
            is StreamResult.Status -> {
                val stage = PipelineStage.fromValue(result.payload.stage)
                if (stage != null) _pipelineStage.value = stage
            }
            is StreamResult.Done -> {
                lastResponseId = result.payload.responseId
            }
            is StreamResult.Error -> {
                _error.value = result.payload.message
                if (result.payload.partial != true) {
                    _streamingBlocks.value = emptyList()
                }
                _isStreaming.value = false
                _pipelineStage.value = PipelineStage.IDLE
            }
        }
    }
}
